templates/problem/voting-controls.html

```diff
@@ -0,0 +1,70 @@
+<div class="vote-form" id="id_vote_form_box">
+    <div>
+        <div id="id_has_voted_prompt">
+            <span class="vote-form-title">{{ _('Change Vote') }}</span>
+            <div id="id_date_box">
+                <span id="id_vote_date"></span>
+            </div>
+        </div>
+        <div id="id_has_not_voted_prompt">
+            <span class="vote-form-title">{{ _('Cast Vote') }}</span>
+        </div>
+        <hr style="margin-bottom: 0px !important;">
+    </div>
+    <form id="id_delete_vote_form">
+        {% csrf_token %}
+    </form>
+    <br>
+    <form id="id_vote_form">
+        {% csrf_token %}
+        <span class="vote-form-info">{{ _('Points:') }}</span>
+        <input id="id_points" class="vote-form-input-text" type="number" step="1"
+               min="{{ min_possible_vote }}" max="{{ max_possible_vote }}" name="points">
+        <br>
+        <div id="id_points_error_box">
+            <br>
+            <span id="id_points_error" class="voting-form-error"></span>
+            <br>
+        </div>
+        <br>
+        <span class="vote-form-info">{{ _('Note:') }}</span>
+        <br>
+        <br>
+        <textarea id="id_note" class="vote-form-input-text" name="note" rows="6" cols="80"></textarea>
+        <br>
+        <div id="id_note_error_box">
+            <br>
+            <span id="id_note_error" class="voting-form-error"></span>
+            <br>
+        </div>
+        <br>
+        <div>
+            <input type="submit" id="id_vote_form_submit_button" class="vote-button" value="{{ _('Vote') }}">
+            <input type="submit" id="id_delete_vote_form_submit_button" value="{{ _('Delete vote') }}" form="id_delete_vote_form">
+        </div>
+    </form>
+</div>
+
+<div>
+    <a class="form-button" id="id_vote_button"></a>
+</div>
+
+{% if vote is not none %}
+    {{ vote.note|json_script('voted_note') }}
+    <script type="text/javascript">
+        set_default_vote_data({{ vote.points }}, new Date('{{ vote.vote_time.isoformat() }}'));
+    </script>
+{% endif %}
+
+<script type="text/javascript">
+    let translated_strings = {
+        'change_vote': '{{ _('Change vote') }}',
+        'vote_points': '{{ _('Vote on problem points') }}',
+        'note_justification_help_text': "{{ _("A short justification for your vote to this problem's point value (Optional).") }}",
+        'last_voted': '{{ _('Last voted on') }}',
```

Comment:
I prefer `'Last voted on {date}'` and `.replace('{date}', 123)`. There are examples of this in the codebase:

https://github.com/DMOJ/online-judge/blob/8b295918ef89d76eb12891841908d4c21a1522a3/templates/comments/media-js.html#L64

In general, sentence fragments such as `'Last voted on'` are bad for i18n.

Reply:
fixed

Reply:
fixed